### PR TITLE
Add step by step nav GA4 tracking

### PR DESCRIPTION
**Pull request by AshGDS**
Hi @andysellick & @JamesCGDS - would you be able to approve this if everything is OK?

Thanks :+1:

## What

- Adds GTM tracking to the step by step nav component. It uses the same GTM object as the accordion tracking but with `type: step by step` instead.
- It is enabled by adding `ga4_tracking: true` when the component is rendered on a page.
- When `ga4_tracking: true` exists, it adds `data-module="ga4-event-tracker" data-ga4-expandable`.  When this module is initialised, any child element that has `data-ga4-event` on it gets tracked on click. Therefore we add `data-ga4-event` where needed, which holds the JSON data we want to send to GTM on click. `data-ga4-expandable` is also added as this signals the `ga4-event-tracker` to track the value of `aria-expanded` and add it to the GTM object as 'opened'/'closed'.
- The Javascript detects the `ga4-event-tracker` module, and populates the GTM JSONs for the 'Show all steps' button, and all the step buttons.

## Why

- This was meant to be added as part of the accordion tracking but we didn't have any confirmation on what the GTM Object should look like. As the step by step nav is very similar to the accordion component, we have used the same GTM object but with `type: "step by step"` instead.

## Visual Changes

None.

### CODE REVIEW

File: app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js
Comment (JamesCGDS) [inline app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js:87]: As per a discussion between @AshGDS and myself, a test will be added to make sure that the ga4 data attributes are being removed and applied correctly

File: CHANGELOG.md
Comment (andysellick) [inline CHANGELOG.md:12]: If a PR only contains one commit I'd generally include the changelog update in the same commit.

File: app/views/govuk_publishing_components/components/_step_by_step_nav.html.erb
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_step_by_step_nav.html.erb:86]: Note: depending on the outcome of https://github.com/alphagov/govuk_publishing_components/pull/3057 and when it gets merged, this might need to be updated to `data-ga4-event`.

File: app/views/govuk_publishing_components/components/docs/step_by_step_nav.yml
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/step_by_step_nav.yml:77]: Are these lines meant to be separate? You need to put a blank line between them for them to render as separate paragraphs in the component guide.

**Review by andysellick — COMMENTED** [review]
Looks good, would definitely echo the comment that this needs some tests. I think a `rb` test to check the right attributes are applied in the right places, and a `js` test to check that the right things happen with the addition of attributes to the show/hide all button.

File: app/views/govuk_publishing_components/components/docs/step_by_step_nav.yml
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/docs/step_by_step_nav.yml:77]: Nope, it was just to make it a bit easier to read in the code

**Comment by AshGDS** [comment]
Thanks @andysellick @JamesCGDS - I've added a JS test and a Ruby test. I'll squash the commits once everything is approved.

@andysellick we don't have ruby tests on the accordions/tabs/header etc. If we're adding a ruby test here, we should add them on the other components too in my opinion.

File: spec/javascripts/components/step-by-step-nav-spec.js
Comment (AshGDS) [inline spec/javascripts/components/step-by-step-nav-spec.js:7]: I'm using `element` here as I'm not using jQuery like the other tests, and variables starting with `$` indicate they're a jQuery variable, so I didn't use it.

File: app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js:179]: @andysellick, yep, we're waiting for that branch to be merged first

**Comment by AshGDS** [comment]
Hi @andysellick - the implementation has changed since James reviewed this. Due to the outcomes of our accordion refactor, I've refactored this implementation. Now the GTM JSON for everything is created using JavaScript. There was no use in hardcoding them onto the `.erb.html` file, as the JavaScript reconstructs the HTML anyway.

I haven't shared the function which adds the GTM JSON to the steps with the accordion function (line 86 in `step-by-step-nav.js` and line 61 in `accordion.js`.) This is because `ga4-core.js` doesn't initialise in the `govuk_publishing_components` repo, so the components will break in our developer frontend if the functions are shared. I think the functions are simple enough that having them duplicated isn't a massive issue.

Would you be able to review the code? Thanks :+1:

File: spec/javascripts/components/step-by-step-nav-spec.js
Comment (AshGDS) [reply spec/javascripts/components/step-by-step-nav-spec.js:938]: @andysellick In fact, `data-ga4-expandable` didn't even need to be added in the JS as it is added by the Ruby template, so I've removed it from the `ga4Data` JS variable.

File: app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js:194]: Nitpick - I'd be tempted to move this line inside the `if` statement above for clarity i.e.

```JavaScript
var ga4Data = ''
// Construct GA4 data-attributes for ga4-event-tracker
if (this.$module.isGa4Enabled) {
  var ga4JSON = {
    event_name: 'select_content',
    type: 'step by step',
    text: titleText.trim(),
    index: i + 1,
    index_total: this.$module.totalSteps
  }
  ga4Data = "data-ga4-event='" + JSON.stringify(ga4JSON) + "'" : ''
}
```

File: app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/components/step-by-step-nav.js:88]: Nitpick - might as well put `this.$module.totalSteps` directly into the object rather than declaring it separately.

**Review by andysellick — APPROVED** [review]
Sorry, a couple more nitpicks. But otherwise good 👍

**Pull request merged**
